add key-value file!

ruclz opened the pull request:
Oceanbase sort needs key-value struct. But the avx512-64bits-qsort only supports key sort. We need key-value sort in this lib.

xtangxtang commented:
The PR is to meet the requirement from Oceanbase.

Oceanbase is a Key/Value store Database. It uses std::sort to sort the object during creating index or other jobs. Not like just sort the key,  Oceanbase passes inner ObStore class object as an array to do the sort job. And the std::sort callback comparing function to determine how to get the key from the ObStore to compare. So in Oceanbase, it swaps the ObStore object index in the array according to the sort function.

Since the avx512 sort in this project do a wondful work in bitonic sort, we want to reuse the core code. But the sort API just do the sort work, doesn't return the indexes of the key after sorted, it is hard for us to integrated the code into Oceanbase. So we create the PR to modify the sort API from _**void avx512_qsort(uint16_t *arr, int64_t arrsize)**_ to _**void avx512_qsort(uint16_t *arr,uint64_t *indexes, int64_t arrsize')**_.  The returned **_indexes_** is used to record the sorted key indexes and help API caller to swap the its objec array outside.

Currently we have created a PR in Oceanbase to merge this project code as a 3rd party lib, now is wating for this PR approval.

r-devulap commented on .vscode/settings.json:
```diff
@@ -0,0 +1,57 @@
+{
```
Please remove this file.

r-devulap commented on Makefile:
```diff
@@ -15,10 +15,10 @@ LD_FLAGS	= -L /usr/local/lib -l $(GTEST_LIB) -l pthread
 
 $(TESTDIR)/%.o : $(TESTDIR)/%.cpp $(SRCS)
-		$(CXX) -march=icelake-client -O3 $(CXXFLAGS) -c $< -o $@
+		$(CXX) -march=icelake-client -O3 $(CXXFLAGS) -l $(GTEST_LIB) -c $< -o $@
```
Why is this needed? This is only building object files and doesn't need any linking to done.

r-devulap commented on Makefile:
```diff
@@ -15,10 +15,10 @@ LD_FLAGS	= -L /usr/local/lib -l $(GTEST_LIB) -l pthread
 
 test: $(TESTDIR)/main.cpp $(TESTOBJS) $(SRCS)
-		$(CXX) tests/main.cpp $(TESTOBJS) $(CXXFLAGS) $(LD_FLAGS) -o testexe
+		$(CXX) tests/main.cpp $(TESTOBJS) $(CXXFLAGS) $(LD_FLAGS) -o testexe 
```
Unnecessary whitespace.

r-devulap reviewed (changes requested):
As far as I understood, you are essentially adding argsort. The indices you pass is also sorted along with the array. Since this is a new feature, could you add new unit tests that cover this feature? 

In general, the PR needs a lot of formatting changes. The diff is pretty hard to read. Could you fix the formatting and run clang-format on the files and re-submit? The formatting specifications are in a _clang-format file.

r-devulap commented:
On a second thought, argsort seems like a special case. You are sorting both key and value arrays.

ruclz commented:
> As far as I understood, you are essentially adding argsort. The modified indices would sort the original array. Correct? Since this is a new feature, could you add new unit tests that cover this feature?
> 
> In general, the PR needs a lot of formatting changes. The diff is pretty hard to read. Could you fix the formatting and run clang-format on the files and re-submit? The formatting specifications are in a _clang-format file.

![image](https://user-images.githubusercontent.com/46258124/215300973-612c7ac8-21b5-4bc5-a72a-98723189ff41.png)

Hello, I run the clang_format execution in the project, but I got the error like the figure, can you have some suggests to solve it? Also, the unit tests can't run on my environment including the unit tests  that you have completed, can you share  your environment to me, like OS, compiler version and so on.

ruclz commented:
> On a second thought, argsort seems like a special case. You are sorting both key and value arrays.

OK, we are sorting key, and the value column exchange with each other according key.

r-devulap commented:
> Hello, I run the clang_format execution in the project, but I got the error like the figure, can you have some suggests to solve it? 

Hmm, not sure why you see that error. I am able to run clang-format on your patch just fine. 

> Also, the unit tests can't run on my environment including the unit tests that you have completed, can you share your environment to me, like OS, compiler version and so on.

Do you see an error? I am using ubuntu latest, g++-10 and running on a SKX. If your processor doesnt support AVX-512, you could run on an [SDE ](https://www.intel.com/content/www/us/en/developer/articles/tool/software-development-emulator.html) too with this command: `sde-external-9.14.0-2022-10-25-lin/sde64 -icx -- ./testexe`

ruclz commented:
OK, I added a unittest for key-value sort, and I run clang_format on all files that I modified. Please help to review. Thanks.

r-devulap commented:
Hi @ruclz This PR modifies the existing `avx512_qsort<T>()` function signature by adding another argument to the function and also adds unnecessary code if someone wants to use only data sorting and not key-value sort. I think a better option would be to move your code to new header files `avx512_64bit_keyvaluesort.hpp` and `avx512-common-keyvaluesort.h `. I understand there might be some code duplication but we can work on minimizing that in a separate PR.

r-devulap reviewed (changes requested):
Move code to separate files and add new API's `avx512_keyvaluesort<T>(`).

ruclz commented:
Hi, @r-devulap, I have moved my code to new header files avx512-64bit-keyvaluesort.hpp and avx512-common-keyvaluesort.h. The unit test can run correctly, please help to review, thanks.

ruclz commented:
Sorry, I forgot to merge the test_all.cpp in the last commit.

r-devulap commented on src/avx512-64bit-keyvaluesort.hpp:
```diff
@@ -0,0 +1,1276 @@
+#define NETWORK_64BIT_4 3, 2, 1, 0, 7, 6, 5, 4
+
+template <>
+struct zmm_kv_vector<int64_t> {
```
Is there a difference between `zmm_kv_vector `struct and `zmm_vector `struct in `avx512-64bit-qsort.hpp`? If not, we should move all these to a new header file `avx512-64bit-common.h` and not duplicate them. You could include the NETWORK_64BIT_* constants too.

r-devulap commented on src/avx512-64bit-keyvaluesort.hpp:
```diff
@@ -0,0 +1,1276 @@
+     * Resort to std::sort if quicksort isnt making any progress
+     */
+    if (max_iters <= 0) {
+        heap_sort<vtype>(keys + left, indexes + left, right - left + 1);
```
why not use std::sort? Just like how you used it in test_all.cpp.

ruclz replied:
Because the arguments are two column pointer(keys and values), if we want to use std::sort, the input of std::sort is a struct, not two arguments, we need create a new two-dimension vector to store key and index, it need to malloc extra space.

r-devulap commented on src/avx512-64bit-keyvaluesort.hpp:
```diff
@@ -0,0 +1,1276 @@
+        qsort_64bit_<vtype>(keys, indexes, pivot_index, right, max_iters - 1);
+}
+
+X86_SIMD_SORT_INLINE int64_t replace_nan_with_inf_kv(double *arr,
```
This is the same function as in `avx512-64bit-qsort.hpp`. You could move this to the common header file as well and no need to re-write it.

r-devulap commented on src/avx512-64bit-keyvaluesort.hpp:
```diff
@@ -0,0 +1,1276 @@
+}
+
+X86_SIMD_SORT_INLINE void
+replace_inf_with_nan_kv(double *arr, int64_t arrsize, int64_t nan_count)
```
same as above. Move this to common header file and do not duplicate.

r-devulap commented on src/avx512-common-keyvaluesort.h:
```diff
@@ -0,0 +1,291 @@
+#else
+#define X86_SIMD_SORT_INLINE static
+#define X86_SIMD_SORT_FINLINE static
+#endif
```
lines 36-82 is duplicated. Move them to new header file and include that here and `avx512-common-qsort.h`

r-devulap commented on src/avx512-16bit-qsort.hpp:
```diff
@@ -686,7 +686,7 @@ replace_inf_with_nan(uint16_t *arr, int64_t arrsize, int64_t nan_count)
 
 template <>
-void avx512_qsort(int16_t *arr, int64_t arrsize)
+inline void avx512_qsort(int16_t *arr, int64_t arrsize)
```
why `inline `these?

r-devulap commented on src/avx512-16bit-qsort.hpp:
```diff
@@ -695,7 +695,7 @@ void avx512_qsort(int16_t *arr, int64_t arrsize)
 
 template <>
-void avx512_qsort(uint16_t *arr, int64_t arrsize)
+inline void avx512_qsort(uint16_t *arr, int64_t arrsize)
```
same. get rid of inline please.

r-devulap commented on src/avx512-32bit-qsort.hpp:
```diff
@@ -682,7 +682,7 @@ replace_inf_with_nan(float *arr, int64_t arrsize, int64_t nan_count)
 
 template <>
-void avx512_qsort<int32_t>(int32_t *arr, int64_t arrsize)
+inline void avx512_qsort<int32_t>(int32_t *arr, int64_t arrsize)
```
remove inline.

r-devulap commented on src/avx512-32bit-qsort.hpp:
```diff
@@ -691,7 +691,7 @@ void avx512_qsort<int32_t>(int32_t *arr, int64_t arrsize)
 
 template <>
-void avx512_qsort<uint32_t>(uint32_t *arr, int64_t arrsize)
+inline void avx512_qsort<uint32_t>(uint32_t *arr, int64_t arrsize)
```
ditto

r-devulap commented on src/avx512-32bit-qsort.hpp:
```diff
@@ -700,7 +700,7 @@ void avx512_qsort<uint32_t>(uint32_t *arr, int64_t arrsize)
 
 template <>
-void avx512_qsort<float>(float *arr, int64_t arrsize)
+inline void avx512_qsort<float>(float *arr, int64_t arrsize)
```
ditto

r-devulap commented on src/avx512-64bit-keyvaluesort.hpp:
```diff
@@ -0,0 +1,1276 @@
+
+template <>
+inline void
+avx512_qsort_kv<int64_t>(int64_t *keys, uint64_t *indexes, int64_t arrsize)
```
Are indices always expected to be a `uint64_t `?

r-devulap reviewed (changes requested):
Still working on reviewing the code, meanwhile lets clean up all the duplicated code.

ruclz commented:
> @ruclz Could you add benchmarking to know how fast this is relative to using std::sort? From issue #8:
> 
> > avx512_qsort_kv(double *keys, uint64_t *indexes, int64_t arrsize) is a little worse than my simple bubble sorting.

OK, I am coding for benchmarking.

ruclz commented:
> @ruclz Could you add benchmarking to know how fast this is relative to using std::sort? From issue #8:
> 
> > avx512_qsort_kv(double *keys, uint64_t *indexes, int64_t arrsize) is a little worse than my simple bubble sorting.

![image](https://user-images.githubusercontent.com/46258124/222083346-3e36c6f0-dc7e-4944-842d-e05baf8b0384.png)

The figure shows the performance of key-value bitonic sort and std::sort. There is no performance drop compared with std::sort.

r-devulap commented on benchmarks/bench.hpp:
```diff
@@ -72,3 +79,49 @@ std::tuple<uint64_t, uint64_t> bench_sort(const std::vector<T> arr,
+
+    for (uint64_t ii = 0; ii < iters; ++ii) {
+        start = cycles_start();
+        std::sort(sortedaar_bckup.begin(),
```
Not sure if std::sort is a fair comparison. The way the key-values are arranged in memory for `std::vector<sorted_t<K, V>>` is very different when compared two distinct vectors  `std::vector<K>, std::vector<V>`. I would think this can alter performance significantly because you can no longer load continuous chunk of memory into ZMM registers. Does Oceanbase have two separate vectors for key and values (I assume they do cos that's the way you implemented the sort). How does Oceanbase currently sort them?

xtangxtang replied:
ObStoreRow is the base unit to do the sort in Oceanbase. The key and value are all ObObject class type in the ObStoreRow. There is no sperate vector for key and value. The ObStoreRow is passed to the std::sort in Oceanbase. std::sort first get the key type from ObObject and then do the sort work.

xtangxtang replied:
To enable the AVX512 sort, we need extract the key from the ObStoreRow and then pass the key and ObStoreRow as the value to avx512_sort(key, value) like API. That's way we need the key-value interface

r-devulap commented on benchmarks/bench-tgl.out:
```diff
@@ -1,28 +1,29 @@
-| uniform random  | int16_t     | 100000     | 1442726         | 19705242  | 13.7     |
-| uniform random  | int16_t     | 1000000    | 20210224        | 212137465 | 10.5     |
-|-----------------+-------------+------------+-----------------+-----------+----------|
+| -----------------+-------------+------------+-----------------+-----------+---------- |
```
please dont check in this file.

r-devulap commented on src/avx512-64bit-keyvaluesort.hpp:
```diff
@@ -0,0 +1,909 @@
+    zmm_t sort;
+    index_vec = zmm_vector<uint64_t>::template i64gather<sizeof(uint64_t)>(
+            rand_index, indexes);
+    sort = sort_zmm_64bit<vtype>(key_vec, index_vec);
```
pivot is picked just based on keys and not index, right? `index_vec` is never really used anywhere in this function, which means you do not need a new `get_pivot_64bit` function. You can move this function to `avx512-64bit-common.h` and just call that instead. You will also need to move `sort_zmm_64_bit<vtype>` into the common file. This should eliminate the need to calling gather function twice and save you some cpu cycles.

ruclz commented:
We have moved the get_povit_64bit function and sort_zmm_64bit function to avx512-64bit-common.h, and modified the bench-tgl.out to origin file. Please help to review again, thanks.

r-devulap commented:
@ruclz please fix the author name and get rid of the unnecessary struct and we should be good to go.

ruclz commented:
Yeah, we have fixed the authors name in the last commit.

r-devulap commented:
LGTM, Thanks @ruclz!

r-devulap commented:
Did a little bit of code cleanup, see #15
